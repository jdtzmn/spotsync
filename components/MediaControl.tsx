import React, { Component, CSSProperties } from 'react'

import PlayIcon from 'react-ionicons/lib/IosPlay'
import PauseIcon from 'react-ionicons/lib/IosPause'

const isFn = (fn) => typeof fn === 'function'
const runIfIsFn = (fn, ...args) => {
  if (isFn(fn)) fn(...args)
}

interface MediaControlProps {
  style?: CSSProperties,
  className?: string,
  size?: string,
  color?: string,
  initialState?: string,
  onPlay?: () => void,
  onPause?: () => void,
  onToggle?: () => void,
}

class MediaControl extends Component<MediaControlProps> {
  state = {
    paused: false
  }

  constructor (props) {
    super(props)

    const startsPaused = props.initialState === 'paused'
    this.state = {
      paused: startsPaused
    }

    this.handleClick = this.handleClick.bind(this)
  }

  handleClick (event) {
    // Toggle the paused state
    const { paused } = this.state
    this.setState({ paused: !paused })

    // Emit necessary events
    const { onToggle, onPlay, onPause } = this.props

    // Emit onPlay if paused and onPause if playing
    if (paused) {
      runIfIsFn(onPlay, event)
    } else {
      runIfIsFn(onPause, event)
    }

    // Emit onToggle event either way
    runIfIsFn(onToggle, event)
  }

  render() {
    const { paused } = this.state
    const { size, color, style, className } = this.props

    let icon
    if (paused) {
      icon = (
        <PlayIcon
          fontSize={size}
          color={color}
          onClick={this.handleClick}
        />
      )
    } else {
      icon = (
        <PauseIcon
          fontSize={size}
          color={color}
          onClick={this.handleClick}
        />
      )
    }

    return (
      <div style={style} className={className}>
        {icon}
      </div>
    )
  }
}

export default MediaControl
