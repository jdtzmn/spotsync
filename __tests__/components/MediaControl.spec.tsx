import * as React from 'react'
import { shallow, mount, ShallowWrapper, ReactWrapper } from 'enzyme'
import MediaControl from 'components/MediaControl'

describe('MediaControl', () => {
  let wrapper: ShallowWrapper
  beforeEach(() => {
    wrapper = shallow(<MediaControl />)
  })

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should render with the correct initialState', () => {
    expect(wrapper.state('paused')).toBe(false)

    const testWrapper = shallow(<MediaControl initialState='paused' />)
    expect(testWrapper.state('paused')).toBe(true)
  })
})

describe('mounted MediaControl', () => {
  let wrapper: ReactWrapper
  beforeEach(() => {
    wrapper = mount(<MediaControl />)
  })

  it('should adjust size based on size prop', () => {
    expect(wrapper.find('svg').prop('width')).toBe('22px')

    wrapper.setProps({ size: '50px' })
    expect(wrapper.find('svg').prop('width')).toBe('50px')
  })

  it('should call onToggle when clicked', () => {
    const onToggle = jest.fn()
    wrapper.setProps({ onToggle })

    wrapper.find('svg').simulate('click')
    expect(onToggle).toHaveBeenCalled()
  })

  it('should only call onPlay when paused', () => {
    // set props
    const onPlay = jest.fn()
    const onPause = jest.fn()
    wrapper.setProps({ onPlay, onPause })

    // set state
    wrapper.setState({ paused: true })

    // simulate click
    wrapper.find('svg').simulate('click')

    expect(onPlay).toHaveBeenCalled()
    expect(onPause).not.toHaveBeenCalled()
  })

  it('should only call onPause when playing', () => {
    // set props
    const onPlay = jest.fn()
    const onPause = jest.fn()
    wrapper.setProps({ onPlay, onPause })

    // set state
    wrapper.setState({ paused: false })

    // simulate click
    wrapper.find('svg').simulate('click')

    expect(onPause).toHaveBeenCalled()
    expect(onPlay).not.toHaveBeenCalled()
  })

  it('should not error when props are not provided', () => {
    // simulate click
    wrapper.find('svg').simulate('click')

    // jest will catch an error if one occurs
  })
})
