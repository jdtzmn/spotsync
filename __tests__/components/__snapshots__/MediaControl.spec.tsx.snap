// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MediaControl should render correctly 1`] = `
<div>
  <IosPause
    beat={false}
    color="#000000"
    fontSize="22px"
    onClick={[Function]}
    rotate={false}
    shake={false}
    style={Object {}}
  />
</div>
`;
